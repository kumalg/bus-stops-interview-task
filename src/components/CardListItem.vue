<template>
    <li class="card-list-item">
        <button
            :tabindex="selectable ? 0 : -1"
            class="button"
            :class="{ selectable, active: selectable && selectedItem === item }"
            type="button"
            @click="selectable && selectItem()"
        >
            {{ item }}
        </button>
    </li>
</template>

<script setup lang="ts">
const emit = defineEmits<{
    (e: 'select-item', value: string): void;
}>();

const props = defineProps<{
    selectable: boolean;
    selectedItem: string | null;
    item: string;
}>();

const selectItem = () => {
    emit('select-item', props.item);
};
</script>

<style lang="scss" scoped>
.card-list-item {
    .button {
        display: block;
        width: 100%;
        height: 3.5rem;
        border: none;
        background-color: transparent;
        text-align: left;
        padding: 0 $card-padding;
        color: $color-gray-800;
        font-family: inherit;
        user-select: none;

        @include default-transition('color, background-color');

        &.selectable {
            cursor: pointer;

            &:hover,
            &.hover {
                background-color: $color-gray-300;
            }

            &:active,
            &.active {
                color: $color-accent;
            }
        }
    }
}
</style>
