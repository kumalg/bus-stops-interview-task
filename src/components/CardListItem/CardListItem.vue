<template>
    <li class="card-list-item">
        <button
            v-if="selectable"
            class="item selectable"
            tabindex="0"
            type="button"
            :class="{ active: selectable && selectedItem === item }"
            @click="selectable && selectItem()"
        >
            {{ item }}
        </button>
        <div v-else class="item">
            {{ item }}
        </div>
    </li>
</template>

<script setup lang="ts">
const emit = defineEmits<{
    (e: 'select-item', value: string): void;
}>();

const props = defineProps<{
    selectable: boolean;
    selectedItem: string | null;
    item: string;
}>();

const selectItem = () => {
    emit('select-item', props.item);
};
</script>

<style lang="scss" scoped>
.card-list-item {
    .item {
        display: flex;
        align-items: center;
        width: 100%;
        height: 3.5rem;
        border: none;
        background-color: transparent;
        text-align: left;
        padding: 0 $card-padding;
        color: $color-gray-800;
        font-family: inherit;

        @include default-transition('color, background-color');

        &.selectable {
            cursor: pointer;
            user-select: none;

            &:hover,
            &.hover {
                background-color: $color-gray-300;
            }

            &:active,
            &.active {
                color: $color-accent;
            }
        }
    }
}
</style>
