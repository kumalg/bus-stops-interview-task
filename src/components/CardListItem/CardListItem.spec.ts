import { shallowMount } from '@vue/test-utils';

import CardListItem from './CardListItem.vue';

const EXAMPLE_ITEM_1 = 'example item';
const EXAMPLE_ITEM_2 = 'example item 2';

describe('CardListItem', () => {
    const CardListItemFactory = (props: InstanceType<typeof CardListItem>['$props']) => {
        return shallowMount(CardListItem, {
            props
        });
    };

    it('renders proper html markup', () => {
        const wrapper = CardListItemFactory({
            item: EXAMPLE_ITEM_1,
            selectable: true,
            selectedItem: EXAMPLE_ITEM_2
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it('add selectable class if selectable is true', async () => {
        const wrapper = CardListItemFactory({
            item: EXAMPLE_ITEM_1,
            selectable: false,
            selectedItem: EXAMPLE_ITEM_2
        });

        const buttonEl = wrapper.find('button.button');

        expect(buttonEl.classes()).not.toContain('selectable');
        expect(buttonEl.attributes()['tabindex']).toBe('-1');

        await wrapper.setProps({
            selectable: true
        });

        expect(buttonEl.classes()).toContain('selectable');
        expect(buttonEl.attributes()['tabindex']).toBe('0');
    });

    it('add active class if selectedItem matches item', async () => {
        const wrapper = CardListItemFactory({
            item: EXAMPLE_ITEM_1,
            selectable: false,
            selectedItem: EXAMPLE_ITEM_1
        });

        const buttonEl = wrapper.find('button.button');

        expect(buttonEl.classes()).not.toContain('active');

        await wrapper.setProps({
            selectable: true
        });

        expect(buttonEl.classes()).toContain('active');
    });

    it('triggers event if selectable is true', async () => {
        const wrapper = CardListItemFactory({
            item: EXAMPLE_ITEM_1,
            selectable: false,
            selectedItem: EXAMPLE_ITEM_2
        });

        const buttonEl = wrapper.find('button.button');
        await buttonEl.trigger('click');

        expect(wrapper.emitted('select-item')).toBe(undefined);

        await wrapper.setProps({
            selectable: true
        });

        await buttonEl.trigger('click');
        expect(wrapper.emitted('select-item')?.length).toBe(1);
    });
});
