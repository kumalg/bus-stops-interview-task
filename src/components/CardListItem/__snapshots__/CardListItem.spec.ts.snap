// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardListItem renders proper html markup 1`] = `
<li
  class="card-list-item"
>
  <button
    class="button selectable"
    tabindex="0"
    type="button"
  >
    example item
  </button>
</li>
`;
